plugins/dbus_announce.c

```diff
@@ -103,9 +119,77 @@ static rpmRC send_ts_message(rpmPlugin plugin,
 
     dbcookie = rpmdbCookie(rpmtsGetRdb(ts));
     rpm_tid_t tid = rpmtsGetTid(ts);
+
+    if (!dbus_message_append_args(msg,
+				  DBUS_TYPE_STRING, &dbcookie,
+				  DBUS_TYPE_UINT32, &tid,
+				  DBUS_TYPE_INVALID))
+	goto err;
+
+    if (!dbus_connection_send(state->bus, msg, NULL))
+	goto err;
+
+    if (snprintf(detail_name, sizeof(detail_name), "%sDetails", name) >= sizeof(detail_name)) {
+        rpmlog(RPMLOG_WARNING,
+	       "dbus_announce plugin: Static buffer for details signal name is not large enough, needs at least %u bytes\n",
+	       (unsigned int)(strlen(name) + strlen("Details") + 1));
+        goto err;
+    }
+
+    dbus_message_unref(msg);
+
+    /* it's used in the log below, thus to have proper signal name in there */
+    name = detail_name;
+    msg = dbus_message_new_signal("/org/rpm/Transaction", /* object name */
+				  "org.rpm.Transaction",  /* interface name */
+				  name);                  /* signal name */
+    if (msg == NULL)
+	goto err;
+
+    nElems = rpmtsNElements(ts);
+    array = (char **) malloc((nElems + 1 ) * sizeof(char *));
+    if (array == NULL)
+        goto err;
+
+    array[0] = NULL;
+    for (i = 0; i < nElems; i++) {
+        rpmte te = rpmtsElement(ts, i);
+        char *buff;
+        int buffln;
+        const char *op = "??";
+        const char *nevra = rpmteNEVRA(te);
+        if (nevra == NULL)
+            nevra = "";
+        switch (rpmteType (te)) {
+        case TR_ADDED:
+            op = "added";
+            break;
+        case TR_REMOVED:
+            op = "removed";
+            break;
+        case TR_RPMDB:
+            op = "rpmdb";
+            break;
+        case TR_RESTORED:
+            op = "restored";
+            break;
+        }
+        buffln = strlen(op) + 1 + strlen(nevra) + 1;
+        buff = (char *) malloc(sizeof(char) * buffln);
+        if (buff == NULL)
+            goto err;
+	/* encode as "operation SPACE nevra" */
+	snprintf(buff, buffln, "%s %s", op, nevra);
```

Comment:
Maybe using rpm's own string functions (i.e. `rstrcat()` in this  case) would be better here (although again, doesn't matter too much).

Reply:
if you want, I can do it, though the `snprintf` helps to avoid buffer overflow, which is a good thing, I believe

Reply:
Oh, I should've mentioned `rasprintf()` here instead. It's a wrapper around the GNU extension `asprintf()` and allows you to get rid of the manual malloc and length calculation. Grep for it in rpm's source code to see some usage examples.

I just checked with @pmatilai and indeed, plugins should generally use rpm's helpers. So, if possible, please use those :)

Actually, same goes for the string array here. The rpm "canonical" way of handling string arrays is through the `argv*()` functions. There's also `argvFree()` which replaces the need to write the custom `free_array()` function here.